### PR TITLE
GH-96: interactive joystick

### LINKED ISSUE #96: Make interactive joystick intarface
When user presses buttons on joystick, UI should reflact those actions
Screenshot 2023-04-18 at 10 02 11

### PR DESCRIPTION
**Pull request by alex-waverley**
make interactive joystick UI (via mouse and joystick itself)

Screenshot 2023-05-02 at 16 06 30

Closes: #96

### CODE REVIEW

File: frontend/src/components/joystick/index.js [hunk 1]
```diff
@@ -20,71 +20,180 @@ const useStyles = makeStyles()((theme) => {
...
-    
-    if (target.dataset.button && event.type === "mousedown") {
-      target.classList.add('active')
+    const buttonName = target.dataset.button
+    const isMouseDownEvent = event.type === 'mousedown'
+    if (buttonName in buttons) {
+      setButtons((currentButtonsState) => ({...currentButtonsState, [buttonName]: isMouseDownEvent}))
     } else {
-      target.classList.remove('active')
+      if(buttonName === axisButtonL3 || buttonName === axisButtonR3) {
```
Comment (sskorol) [inline frontend/src/components/joystick/index.js]: reformat this file please, or maybe it's probably better setting up prettier

File: frontend/src/components/joystick/index.js [hunk 2]
```diff
@@ -20,71 +20,180 @@ const useStyles = makeStyles()((theme) => {
...
+
+  const axisMoveHandler = (e) => {
+    if(axisButtonPressed) {
+      const buttonName = e.target.dataset.button
+      let shiftX = e.pageX - axisShift[0]
+      let shiftY = axisShift[1] - e.pageY
+      if (Math.abs(shiftX) > AxisMoveRadius) {
+        shiftX = shiftX < 0 ? -AxisMoveRadius : AxisMoveRadius
+      }
+      if (Math.abs(shiftY) > AxisMoveRadius) {
```
Comment (sskorol) [inline frontend/src/components/joystick/index.js]: let's extract it, as it's the same block of code as above, depending on a single variable value

File: frontend/src/components/joystick/index.js [hunk 3]
```diff
@@ -20,71 +20,180 @@ const useStyles = makeStyles()((theme) => {
...
+      }
+      // calculate distance between center of the circle and current cursor position
+      const d = Math.sqrt(Math.pow(shiftX, 2) + Math.pow(shiftY, 2))
+
+      // if the distance is bigger than radius of the circle we need to find the coordinates of the dot on a circle
+      if (d > AxisMoveRadius) {
+        shiftX = shiftX / d * AxisMoveRadius
+        shiftY = shiftY / d * AxisMoveRadius
+      }
+      setAxis((currentAxisState) => ({
```
Comment (sskorol) [inline frontend/src/components/joystick/index.js:143]: seems pretty similar to what's defined in a `handleClickEvent`, but with a difference in a right part value

**Review by sskorol — CHANGES_REQUESTED** [review]

File: frontend/src/components/joystick/index.js (hunk 1 above)
Comment (alex-waverley) [reply frontend/src/components/joystick/index.js]: done

File: frontend/src/components/joystick/index.js (hunk 2 above)
Comment (alex-waverley) [reply frontend/src/components/joystick/index.js]: done

File: frontend/src/components/joystick/index.js (hunk 3 above)
Comment (alex-waverley) [reply frontend/src/components/joystick/index.js:143]: done

**Review by sskorol — APPROVED** [review]

**Pull request merged**